cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaReconciler.java

```diff
@@ -483,6 +487,67 @@ protected Future<Void> initClusterRoleBinding() {
         ).map((Void) null);
     }
 
+    protected Future<Set<String>> topicNames(Admin ac) {
+
+        Promise<Set<String>> namesPromise = Promise.promise();
+        ac.listTopics(new ListTopicsOptions().listInternal(true)).names()
+                .whenComplete((names, error) -> {
+                    if (error != null) {
+                        namesPromise.fail(error);
+                    } else {
+                        LOGGER.debugCr(reconciliation, "Got {} topic names", names.size());
+                        namesPromise.complete(names);
+                    }
+                });
+        return namesPromise.future();
+    }
+
+    protected Future<Collection<TopicDescription>> describeTopics(Admin ac, Set<String> names) {
+        Promise<Collection<TopicDescription>> descPromise = Promise.promise();
+        ac.describeTopics(names).allTopicNames()
```

Comment:
Here you are describing all topics in the cluster. If there are a lot of partitions this could be a very large response. Bear in mind that with KRaft Kafka should be able to scale to ~1M partitions in a cluster. While Strimzi might not want to go _that_ far it's still worth thinking about how the algorithm you're using here currently requires to have all the `TopicDescriptions` in memory at once. It doesn't need to. You could split up the `names` into batches, do `describeTopics` for each batch and then do the computation in `groupTopicsByBroker` for each batch, stopping on the first batch where the broker has replicas. 

(There is a bit of a trade off in doing this, since it means you'd make multiple `describeTopics` calls over time (and they'll be handled by different brokers), which means there's the possibility of a race, where you find that the broker has no replicas and proceed to scale down, but the broker gets assigned replicas in the meantime. However, this race exists anyway, even in with the code you have now. The only difference is how large the time window is. I'm inclined to think it's better to opt for the scalability over a slightly smaller window.)